tools/osbuild-depsolve-dnf

```diff
@@ -17,11 +19,485 @@ import urllib.request
+    from dnf.exceptions import DepsolveError, MarkingError, RepoError
+    from dnf.exceptions import Error as DnfError
+    solverClassName = "SolverDNF"
+elif argv0 == "dnf5":
+    import libdnf5 as dnf5
+    from libdnf5.base import GoalProblem_NO_PROBLEM as NO_PROBLEM
+    from libdnf5.common import QueryCmp_CONTAINS as CONTAINS
+    from libdnf5.common import QueryCmp_EQ as EQ
+    from libdnf5.common import QueryCmp_GLOB as GLOB
+    solverClassName = "SolverDNF5"
+    # Define here to avoid having to change the common code, those will
+    # never the thrown. With a bit of thinking this can probably be done
+    # more elegantly
+    class MarkingError(Exception):
+        pass
+
+    class DepsolveError(Exception):
+        pass
+
+    class RepoError(Exception):
+        pass
+
+    class DnfError(Exception):
+        pass
+
+
+class Solver(abc.ABC):
+    @abc.abstractmethod
+    def dump(self):
+        pass
+
+    @abc.abstractmethod
+    def depsolve(self, arguments):
+        pass
+
+    @abc.abstractmethod
+    def search(self, args):
+        pass
+
+
+class SolverBase(Solver):
+    # put any shared helpers in here
+    pass
+
+
+# support for dnf5, maybe put under SolverDNF5?
+# XXX - Temporarily lifted from dnf.rpm module  # pylint: disable=fixme
+def _invert(dct):
+    return {v: k for k in dct for v in dct[k]}
+
+
+_BASEARCH_MAP = _invert({
+    'aarch64': ('aarch64',),
+    'alpha': ('alpha', 'alphaev4', 'alphaev45', 'alphaev5', 'alphaev56',
+              'alphaev6', 'alphaev67', 'alphaev68', 'alphaev7', 'alphapca56'),
+    'arm': ('armv5tejl', 'armv5tel', 'armv5tl', 'armv6l', 'armv7l', 'armv8l'),
+    'armhfp': ('armv6hl', 'armv7hl', 'armv7hnl', 'armv8hl'),
+    'i386': ('i386', 'athlon', 'geode', 'i386', 'i486', 'i586', 'i686'),
+    'ia64': ('ia64',),
+    'mips': ('mips',),
+    'mipsel': ('mipsel',),
+    'mips64': ('mips64',),
+    'mips64el': ('mips64el',),
+    'loongarch64': ('loongarch64',),
+    'noarch': ('noarch',),
+    'ppc': ('ppc',),
+    'ppc64': ('ppc64', 'ppc64iseries', 'ppc64p7', 'ppc64pseries'),
+    'ppc64le': ('ppc64le',),
+    'riscv32': ('riscv32',),
+    'riscv64': ('riscv64',),
+    'riscv128': ('riscv128',),
+    's390': ('s390',),
+    's390x': ('s390x',),
+    'sh3': ('sh3',),
+    'sh4': ('sh4', 'sh4a'),
+    'sparc': ('sparc', 'sparc64', 'sparc64v', 'sparcv8', 'sparcv9',
+              'sparcv9v'),
+    'x86_64': ('x86_64', 'amd64', 'ia32e'),
+})
+
+
+def remote_location(package, schemes=("http", "ftp", "file", "https")):
+    """Return the remote url where a package rpm may be downloaded from
+
+    This wraps the get_remote_location() function, returning the first
+    result or if it cannot find a suitable url it raises a RuntimeError
+    """
+    urls = package.get_remote_locations(schemes)
+    if not urls or len(urls) == 0:
+        raise RuntimeError(f"Cannot determine remote location for {package.get_nevra()}")
+
+    return urls[0]
+
+
+class TransactionError(Exception):
```

Comment:
These exception classes kinda get lost down here, I'd move them up before the functions, after the base classes

Reply:
I'm happy to dive into them and rework that part of the code a bit and unify them some more. But I also don't want to overinvest as this change seems very contentious and likely to not land (at least AIUI today).